File: frontend/types/itinerary.ts
/**
 * Itinerary domain types.
 *
 * Mirror the Sprint 2 backend itinerary API (FastAPI + Pydantic).
 * Read shapes match `ItineraryDayRead` / `ItineraryItemRead` /
 * `AgentRunRead`. All fields are snake_case — the backend only uses
 * camelCase aliases internally when validating raw LLM JSON, not at the
 * API boundary.
 *
 * `estimated_cost` is a Pydantic `Decimal`, which serializes to a JSON
 * *string* (e.g. "12.50"), same as `Trip.budget` in `types/trip.ts`.
 */

export type ItineraryItemType =
  | "activity"
  | "meal"
  | "hotel"
  | "transport"
  | "rest"
  | "event"
  | "free_time";

export type WalkingIntensity = "low" | "medium" | "high";

export type ItineraryItemPriority = "required" | "recommended" | "optional";

export type AgentRunStatus = "pending" | "running" | "completed" | "failed";

/** A resolved place as returned by the API (`PlaceRead`). */
export interface Place {
  id: number;
  google_place_id: string;
  name: string;
  address: string | null;
  lat: number | null;
  lng: number | null;
  rating: number | null;
  price_level: number | null;
  types: string[];
  opening_hours: Record<string, unknown> | null;
  maps_url: string;
  created_at: string;
  updated_at: string;
}

/** An itinerary item as returned by the API (`ItineraryItemRead`). */
export interface ItineraryItem {
  id: number;
  itinerary_day_id: number;
  start_time: string;
  end_time: string;
  title: string;
  type: ItineraryItemType;
  location_name: string | null;
  description: string | null;
  estimated_cost: string | null;
  verified_cost: string | null;
  price_source: string | null;
  walking_intensity: WalkingIntensity | null;
  priority: ItineraryItemPriority;
  order_index: number;
  place: Place | null;
  travel_time_to_next_minutes: number | null;
  distance_to_next_meters: number | null;
  travel_mode_to_next: string | null;
}

/** An itinerary day as returned by the API (`ItineraryDayRead`). */
export interface ItineraryDay {
  id: number;
  trip_id: number;
  day_number: number;
  date: string | null;
  theme: string | null;
  summary: string | null;
  total_walking_minutes: number | null;
  total_transit_minutes: number | null;
  total_distance_meters: number | null;
  route_optimized: boolean;
  items: ItineraryItem[];
}

/** Per-day route totals as returned by the API (`RouteDaySummary`). */
export interface RouteDaySummary {
  day_number: number;
  date: string | null;
  theme: string | null;
  route_optimized: boolean;
  total_walking_minutes: number | null;
  total_transit_minutes: number | null;
  total_distance_meters: number | null;
  item_count: number;
}

/** An agent run as returned by the API (`AgentRunRead`). */
export interface AgentRun {
  id: number;
  trip_id: number;
  status: AgentRunStatus;
  model_used: string | null;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

File: backend/app/schemas/itinerary.py
"""Itinerary API and AI-output schemas.

Two families live here:

- ``*Read`` schemas mirror the persisted ORM models for API responses.
- ``*AI`` schemas validate the raw JSON returned by the LLM before it is
  mapped to DB rows. They use camelCase aliases so they match the exact JSON
  shape the model is prompted to return, while exposing snake_case fields to
  Python code (``populate_by_name=True``).
"""

# Imported as a module (not ``from datetime import date``) because the itinerary
# schemas expose a field literally named ``date``, which would otherwise shadow
# the type at class-body evaluation time.
import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.itinerary import (
    ItineraryItemPriority,
    ItineraryItemType,
    WalkingIntensity,
)
from app.schemas.place import PlaceRead


# --- Read schemas (mirror ORM models) ---------------------------------------


class ItineraryItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    itinerary_day_id: int
    start_time: str
    end_time: str
    title: str
    type: ItineraryItemType
    location_name: str | None = None
    description: str | None = None
    estimated_cost: Decimal | None = None
    verified_cost: Decimal | None = None
    price_source: str | None = None
    walking_intensity: WalkingIntensity | None = None
    priority: ItineraryItemPriority
    order_index: int
    place: PlaceRead | None = None
    travel_time_to_next_minutes: int | None = None
    distance_to_next_meters: int | None = None
    travel_mode_to_next: str | None = None
    created_at: datetime.datetime
    updated_at: datetime.datetime


class ItineraryDayRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    trip_id: int
    day_number: int
    date: datetime.date | None = None
    theme: str | None = None
    summary: str | None = None
    total_walking_minutes: int | None = None
    total_transit_minutes: int | None = None
    total_distance_meters: int | None = None
    route_optimized: bool
    created_at: datetime.datetime
    updated_at: datetime.datetime
    items: list[ItineraryItemRead] = Field(default_factory=list)


class RouteDaySummary(BaseModel):
    model_config = ConfigDict()

    day_number: int
    date: str | None
    theme: str | None
    route_optimized: bool
    total_walking_minutes: int | None
    total_transit_minutes: int | None
    total_distance_meters: int | None
    item_count: int


# --- AI-output schemas (validate raw LLM JSON, not persisted) ---------------
#
# The LLM proposes a flat *pool* of places (one hotel, many activities/events,
# many restaurant options) with no day assignment or timing — the backend
# (``app.services.day_planner_service``) clusters them into geographically
# tight days, sequences each day, and assigns time blocks. A second, cheap
# LLM call (``DayNarrationAI``) fills in day themes/summaries afterwards.


class HotelAI(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str
    description: str | None = None
    estimated_cost_per_night: Decimal | None = Field(
        default=None, alias="estimatedCostPerNight"
    )


class ActivityAI(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str
    type: ItineraryItemType  # ACTIVITY or EVENT
    duration_minutes: int = Field(default=90, alias="durationMinutes")
    priority: ItineraryItemPriority
    walking_intensity: WalkingIntensity | None = Field(
        default=None, alias="walkingIntensity"
    )
    description: str | None = None
    estimated_cost: Decimal | None = Field(default=None, alias="estimatedCost")
    best_time_of_day: Literal["morning", "afternoon", "evening", "any"] = Field(
        default="any", alias="bestTimeOfDay"
    )


class RestaurantAI(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str
    meal_type: Literal["lunch", "dinner"] = Field(alias="mealType")
    description: str | None = None
    estimated_cost: Decimal | None = Field(default=None, alias="estimatedCost")


class TripPlanAIResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    trip_title: str = Field(alias="tripTitle")
    overview: str
    hotel: HotelAI
    activities: list[ActivityAI] = Field(default_factory=list)
    restaurants: list[RestaurantAI] = Field(default_factory=list)


class DayNarrationAI(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    day_number: int = Field(alias="dayNumber")
    theme: str
    summary: str


class DayNarrationBatchAI(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    days: list[DayNarrationAI] = Field(default_factory=list)
